### PR TITLE
Add support and tests for URLComponents.percentEncodedQueryItems

### PR DESCRIPTION
**Pull request by rauhul**
rdar://59439703 (URLComponents.percentEncodedQueryItems missing on linux)

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci test linux

File: CoreFoundation/URL.subproj/CFURLComponents.c [hunk 1]
```diff
@@ -1409,6 +1409,9 @@ static Boolean _CFURLComponentsSetQueryItemsInternal(CFURLComponentsRef componen
                 if ( name && name != kCFNull ) {
                     if ( addPercentEncoding ) {
                         CFStringRef stringWithPercentEncoding = _CFStringCreateByAddingPercentEncodingWithAllowedCharacters(kCFAllocatorSystemDefault, name, queryNameValueAllowed);
+                        if ( !stringWithPercentEncoding ) {
```
Comment (millenomi) [inline CoreFoundation/URL.subproj/CFURLComponents.c:1412]: Why the change here vis-a-vis Darwin?

Comment (rauhul) [reply CoreFoundation/URL.subproj/CFURLComponents.c:1412]: I ported this condition from Darwin, following up internally.

File: Sources/Foundation/NSURL.swift [hunk 2]
```diff
@@ -1444,8 +1450,8 @@ open class NSURLComponents: NSObject, NSCopying {
             return (0..<count).map { idx in
                 let oneEntry = unsafeBitCast(CFArrayGetValueAtIndex(queryArray, idx), to: NSDictionary.self)
                 let swiftEntry = oneEntry._swiftObject
-                let entryName = swiftEntry["name"] as! String
-                let entryValue = swiftEntry["value"] as? String
+                let entryName = swiftEntry[_kCFURLComponentsNameKey as! String] as! String
+                let entryValue = swiftEntry[_kCFURLComponentsValueKey as! String] as? String
```
Comment (rauhul) [inline Sources/Foundation/NSURL.swift:1454]: @millenomi could this change be causing the ci failure?
> Fatal error: Constant strings cannot be deallocated: file /home/buildnode/jenkins/workspace/swift-corelibs-foundation-PR-Linux/swift-corelibs-foundation/Sources/Foundation/NSCFString.swift, line 119

Backtrace:
```
0    libswiftCore.so                    0x00007f88d7c7cfb0 swift_reportError + 50
1    libswiftCore.so                    0x00007f88d7cefea0 _swift_stdlib_reportFatalErrorInFile + 115
5    libswiftCore.so                    0x00007f88d79b1270 _assertionFailure(_:_:file:line:flags:) + 523
7    libFoundation.so                   0x00007f88d87aefc0 _NSCFConstantString.__deallocating_deinit + 25
10   libFoundation.so                   0x00007f88d8b089e0 CFRelease + 73
15   libFoundation.so                   0x00007f88d8b09a20 _CFDeinit + 106
16   libFoundation.so                   0x00007f88d87a56d0 _NSCFDictionary.deinit + 47
17   libFoundation.so                   0x00007f88d87a57f0 _NSCFDictionary.__deallocating_deinit + 25
20   libFoundation.so                   0x00007f88d8b089e0 CFRelease + 73
24   libFoundation.so                   0x00007f88d8b09a20 _CFDeinit + 106
25   libFoundation.so                   0x00007f88d879bd40 _NSCFArray.deinit + 47
26   libFoundation.so                   0x00007f88d879be60 _NSCFArray.__deallocating_deinit + 25
28   libFoundation.so                   0x00007f88d89a0630 NSURLComponents.queryItems.getter + 756
32   libFoundation.so                   0x00007f88d84f8730 _MutableHandle.map<A>(_:) + 189
33   libFoundation.so                   0x00007f88d8aec020 URLComponents.queryItems.getter + 120
```

I'm not sure how it would since `_kCFURLComponentsNameKey as! String` should just be used to compute a hash but perhaps I'm misunderstanding something.

**Comment by rauhul** [comment]
@spevans @compnerd @tomerd is there a simple way to get a linux dev environment? Maybe some docker image I can pull that has most things built so I can incremental compile?

**Comment by millenomi** [comment]
The easiest way is to build your own Dockerfile inheriting from an existing Ubuntu file, and have that install the necessary prerequisites per the Swift README.

Unfortunately SCF’s build is tangled enough in the Swift process that it is easier to run the larger build than to guess how to set up the CMake environment just for this one repo.

**Comment by tomerd** [comment]
hi @rauhul the https://github.com/apple/swift-docker/tree/master/swift-ci has the docker images that we use to build & qualify the toolchain with. also see https://github.com/apple/swift-docker/pull/171 for some additional information on how to use them

**Comment by millenomi** [comment]
@rauhul Are you still interested in shepherding this patch?

**Comment by rauhul** [comment]
> @rauhul Are you still interested in shepherding this patch?

I don't think I'm going to have anytime for the next couple months unfortunately :(. I'd be happy to let someone else get this through.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by shahmishal** [comment]
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create pull request with main branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412
